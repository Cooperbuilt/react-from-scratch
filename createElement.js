/**
* create element function
*/

/**
* This creates an object representing a dom node,
* @method render
* @param {string} type this is the element type (i.e "div", "span" etc.)
* @param {Object} props object representing all element props (i.e class, id, href, etc.)
* @param {Object} TODO <= fix this
* @return {object}
*/

const TEXT_ELEMENT = "TEXT ELEMENT";

export const createElement = (type, config, ...rest) => {
  const props = Object.assign({}, config);
  if (rest.length) {
    props.children = rest
    .filter(maybeChild => maybeChild != null && maybeChild !== false)
    .map(child => child instanceof Object ? child : createTextElement(child));  
  }
  return { type, props };
}

const createTextElement = (value) => {
  return createElement(TEXT_ELEMENT, { nodeValue: value });
}