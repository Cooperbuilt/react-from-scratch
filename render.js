/**
 * Render Methods
 */

/**
 * This renders an element to the dom
 * @method render
 * @param {object} element this is the element to render
 * @param {Object} parentDom A dom object
 * @return {void} executes a side effect
 */

export function render(element, parentDom) {
  const { type, props } = element;

  // Create the DOM element
  const isTextElement = type === "TEXT ELEMENT";
  const dom = isTextElement
    ? document.createTextNode("")
    : document.createElement(type);

  // Add event listeners
  const isListener = name => name.startsWith("on");
  Object.keys(props)
    .filter(isListener)
    .forEach(name => {
      const eventType = name.toLowerCase().substring(2);
      dom.addEventListener(eventType, props[name]);
    });

  // Set Props only on HTML elements and skip nodes
  if (dom instanceof HTMLElement) {
    const isAttribute = name => !isListener(name) && name != "children";
    Object.keys(props)
      .filter(isAttribute)
      .forEach(name => {
        dom.setAttribute(name, props[name]);
      });
  }

  // Render children
  const childElements = props.children || [];
  childElements.forEach(childElement => render(childElement, dom));

  // Append to parent
  parentDom.appendChild(dom);
}
