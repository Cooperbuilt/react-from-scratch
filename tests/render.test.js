/**
 * Render Spec
 *
 * @author Evan Cooper
 */
import { render, reconcile, instantiate } from "../render";
import jsdom from "jsdom";
import {createElement} from '../createElement';
const { JSDOM } = jsdom;

const dom = new JSDOM("<!DOCTYPE html><head/><body></body>");
global.window = dom.window;
global.document = window.document;
global.navigator = window.navigator;

const ELEMENT = {
  type: "div",
  props: {
    children: [{ type: "input", props: {} }]
  }
};

const anonJestFn = jest.fn();

const ELEMENT_WITH_PROPS = {
  type: "div",
  props: {
    id: "container",
    children: [
      {
        type: "input",
        props: {
          value: "foo",
          type: "text",
          class: "input",
          onClick: anonJestFn
        }
      },
      {
        type: "a",
        props: {
          href: "/bar",
          children: [
            {
              type: "TEXT ELEMENT",
              props: { nodeValue: "Foo" }
            }
          ]
        }
      }
    ]
  }
};
const ELEMENT_WITH_CUSTOM_PROPS = {
  type: "div",
  props: {
    id: "container",
    children: [
      {
        type: "input",
        props: { "data-enzyme-id": "inputElement", type: "text" }
      }
    ]
  }
};

describe("Render function", () => {
  let parentNode;
  beforeAll(() => {
    let div = document.createElement("div");
    div.setAttribute("class", "root");
    document.body.appendChild(div);
    parentNode = document.querySelector(".root");
  });

  it("appends an element, including it's children, to the dom", () => {
    render(ELEMENT, parentNode);
    expect(document.querySelector("div").innerHTML).toEqual(
      "<div><input></div>"
    );
  });
  it("appends an element and adds props, listeners, and text nodes", () => {
    render(ELEMENT_WITH_PROPS, parentNode);
    document.querySelector(".input").click();
    expect(anonJestFn).toHaveBeenCalled();
    expect(document.querySelector("div").innerHTML).toBe(
      '<div><input></div><div id="container"><input value="foo" type="text" class="input"><a href="/bar">Foo</a></div>'
    );
  });
});

describe("instantiate function", () => {
  it("Creates an instance of an element that contains dom, element, and childInstances if provided", () => {
    expect(instantiate(ELEMENT)).toMatchSnapshot()
  });
});
