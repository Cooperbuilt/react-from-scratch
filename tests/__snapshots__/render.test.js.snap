// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`instantiate function Creates an instance of an element that contains dom, element, and childInstances if provided 1`] = `
Object {
  "childInstances": Array [
    Object {
      "childInstances": Array [],
      "dom": <input />,
      "element": Object {
        "props": Object {},
        "type": "input",
      },
    },
  ],
  "dom": <div>
    <input />
  </div>,
  "element": Object {
    "props": Object {
      "children": Array [
        Object {
          "props": Object {},
          "type": "input",
        },
      ],
    },
    "type": "div",
  },
}
`;
